### PR TITLE
Esdt set all function

### PR DESCRIPTION
**Pull request by sasurobert**
Esdt set all function

### CODE REVIEW

File: vm/systemSmartContracts/esdt.go [hunk 1]
```diff
@@ -1326,82 +1440,110 @@ func (e *esdt) checkSpecialRolesAccordingToTokenType(args [][]byte, token *ESDTD
...
-
-	token, returnCode := e.basicOwnershipChecks(args)
-	if returnCode != vmcommon.Ok {
-		return returnCode
-	}
+func (e *esdt) setRolesForTokenAndAddress(
+	token *ESDTDataV2,
+	address []byte,
+	roles [][]byte,
+) (bool, bool, vmcommon.ReturnCode) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:1447]: this is prone to programming errors. Strong suggestion to use a simple, un-exported DTO

File: vm/systemSmartContracts/esdt.go [hunk 2]
```diff
@@ -66,6 +66,8 @@ type esdt struct {
 	flagNFTCreateONMultiShard        atomic.Flag
 	metaESDTEnableEpoch              uint32
 	flagMetaESDT                     atomic.Flag
+	registerAndSetEnableEpoch        uint32
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:69]: registerAndSetAllEnableEpoch or registerAndSetRolesEnableEpoch?

File: vm/systemSmartContracts/esdt.go [hunk 3]
```diff
@@ -66,6 +66,8 @@ type esdt struct {
...
+	flagRegisterAndSet               atomic.Flag
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:70]: flagRegisterAndSetAll or flagRegisterAndSetRoles?

File: config/epochConfig.go [hunk 4]
```diff
@@ -62,6 +62,7 @@ type EnableEpochs struct {
 	RemoveNonUpdatedStorageEnableEpoch          uint32
 	DeleteDelegatorAfterClaimRewardsEnableEpoch uint32
 	OptimizeNFTStoreEnableEpoch                 uint32
+	ESDTRegisterAndSetAllEnableEpoch            uint32
```
Comment (SebastianMarian) [inline config/epochConfig.go:65]: ESDTRegisterAndSetRolesEnableEpoch ?

File: vm/systemSmartContracts/esdt.go [hunk 5]
```diff
@@ -166,6 +170,8 @@ func (e *esdt) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
 		return e.registerMetaESDT(args)
 	case "changeSFTToMetaESDT":
 		return e.changeSFTToMetaESDT(args)
+	case "registerAndSetAll":
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:173]: registerAndSetRoles ?

File: vm/systemSmartContracts/esdt.go [hunk 6]
```diff
@@ -447,6 +453,114 @@ func (e *esdt) registerMetaESDT(args *vmcommon.ContractCallInput) vmcommon.Retur
...
+	e.eei.Finish(tokenIdentifier)
+	e.eei.AddLogEntry(logEntry)
+
+	return vmcommon.Ok
+}
+
+func getAllRolesForTokenType(tokenType string) ([][]byte, error) {
+	switch tokenType {
+	case core.NonFungibleESDT:
+		return [][]byte{[]byte(core.ESDTRoleNFTCreate), []byte(core.ESDTRoleNFTBurn), []byte(core.ESDTRoleNFTUpdateAttributes), []byte(core.ESDTRoleNFTAddURI)}, nil
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:540]: I would leave it here. and would use as little global vars as possible. as the definition for this list is not a constant.

File: vm/systemSmartContracts/esdt.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1447]: done

File: vm/systemSmartContracts/esdt.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:70]: done

File: config/epochConfig.go (hunk 4 above)
Comment (sasurobert) [reply config/epochConfig.go]: done

File: vm/systemSmartContracts/esdt.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:173]: done

File: vm/systemSmartContracts/esdt.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:69]: done

File: vm/systemSmartContracts/esdt.go [hunk 7]
```diff
@@ -507,12 +507,12 @@ func (e *esdt) registerAndSetRoles(args *vmcommon.ContractCallInput) vmcommon.Re
...
 	}
 
-	transferRoleExists, isMultiShardNFTCreateSet, returnCode := e.setRolesForTokenAndAddress(token, args.CallerAddr, allRoles)
+	properties, returnCode := e.setRolesForTokenAndAddress(token, args.CallerAddr, allRoles)
 	if returnCode != vmcommon.Ok {
 		return returnCode
 	}
 
-	returnCode = e.prepareAndSendRoleChangeData(tokenIdentifier, args.CallerAddr, allRoles, transferRoleExists, isMultiShardNFTCreateSet)
+	returnCode = e.prepareAndSendRoleChangeData(tokenIdentifier, args.CallerAddr, allRoles, properties.transferRoleExists, properties.isMultiShardNFTCreateSet)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:515]: could have passed directly the properties struct

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -1528,12 +1537,12 @@ func (e *esdt) setSpecialRole(args *vmcommon.ContractCallInput) vmcommon.ReturnC
...
 	}
 
-	transferRoleExists, isMultiShardNFTCreateSet, returnCode := e.setRolesForTokenAndAddress(token, args.Arguments[1], args.Arguments[2:])
+	properties, returnCode := e.setRolesForTokenAndAddress(token, args.Arguments[1], args.Arguments[2:])
 	if returnCode != vmcommon.Ok {
 		return returnCode
 	}
 
-	returnCode = e.prepareAndSendRoleChangeData(args.Arguments[0], args.Arguments[1], args.Arguments[2:], transferRoleExists, isMultiShardNFTCreateSet)
+	returnCode = e.prepareAndSendRoleChangeData(args.Arguments[0], args.Arguments[1], args.Arguments[2:], properties.transferRoleExists, properties.isMultiShardNFTCreateSet)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:1545]: could have passed directly the properties struct

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — DISMISSED** [review]

File: node/nodeRunner.go [hunk 9]
```diff
@@ -161,7 +161,7 @@ func printEnableEpochs(configs *config.Configs) {
 	log.Debug(readEpochFor("delete delegator data after claim rewards"), "epoch", enableEpochs.DeleteDelegatorAfterClaimRewardsEnableEpoch)
 	log.Debug(readEpochFor("optimize nft metadata store"), "epoch", enableEpochs.OptimizeNFTStoreEnableEpoch)
 	log.Debug(readEpochFor("create nft through execute on destination by caller"), "epoch", enableEpochs.CreateNFTThroughExecByCallerEnableEpoch)
-	log.Debug(readEpochFor("esdt: enable epoch for esdt register and set all function"), "epoch", enableEpochs.ESDTRegisterAndSetAllEnableEpoch)
+	log.Debug(readEpochFor("esdt: enable epoch for esdt register and set all function"), "epoch", enableEpochs.ESDTRegisterAndSetAllRolesEnableEpoch)
```
Comment (SebastianMarian) [inline node/nodeRunner.go:164]: and set all roles

File: vm/systemSmartContracts/esdt.go [hunk 10]
```diff
@@ -116,25 +116,25 @@ func NewESDTSmartContract(args ArgsNewESDTSmartContract) (*esdt, error) {
...
+		endOfEpochSCAddress:               args.EndOfEpochSCAddress,
+		addressPubKeyConverter:            args.AddressPubKeyConverter,
 	}
 	log.Debug("esdt: enable epoch for esdt", "epoch", e.enabledEpoch)
 	log.Debug("esdt: enable epoch for contract global mint and burn", "epoch", e.globalMintBurnDisableEpoch)
 	log.Debug("esdt: enable epoch for contract transfer role", "epoch", e.transferRoleEnableEpoch)
 	log.Debug("esdt: enable epoch for esdt NFT create on multiple shards", "epoch", e.nftCreateONMultiShardEnableEpoch)
 	log.Debug("esdt: enable epoch for meta tokens, financial SFTs", "epoch", e.metaESDTEnableEpoch)
-	log.Debug("esdt: enable epoch for esdt register and set all function", "epoch", e.registerAndSetEnableEpoch)
+	log.Debug("esdt: enable epoch for esdt register and set all function", "epoch", e.registerAndSetAllRolesEnableEpoch)
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:137]: and set all roles

File: vm/systemSmartContracts/esdt.go [hunk 11]
```diff
@@ -1974,8 +1983,8 @@ func (e *esdt) EpochConfirmed(epoch uint32, _ uint64) {
 	e.flagMetaESDT.Toggle(epoch >= e.metaESDTEnableEpoch)
 	log.Debug("ESDT contract financial SFTs", "enabled", e.flagMetaESDT.IsSet())
 
-	e.flagRegisterAndSet.Toggle(epoch >= e.registerAndSetEnableEpoch)
-	log.Debug("ESDT register and set all", "enabled", e.flagRegisterAndSet.IsSet())
+	e.flagRegisterAndSetAllRoles.Toggle(epoch >= e.registerAndSetAllRolesEnableEpoch)
+	log.Debug("ESDT register and set all", "enabled", e.flagRegisterAndSetAllRoles.IsSet())
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/esdt.go:1986]: and set all roles

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/esdt.go (hunk 10 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:137]: done

File: vm/systemSmartContracts/esdt.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1986]: done

**Review by iulianpascalau — DISMISSED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.

**Pull request merged**